src/GafferScene/MotionPath.cpp

```diff
@@ -0,0 +1,380 @@
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/MotionPath.h"
+
+#include "GafferScene/Isolate.h"
+#include "GafferScene/SceneAlgo.h"
+
+#include "Gaffer/Context.h"
+#include "Gaffer/NumericPlug.h"
+
+#include "IECoreScene/CurvesPrimitive.h"
+
+#include "OpenEXR/ImathMatrixAlgo.h"
+
+using namespace IECore;
+using namespace IECoreScene;
+using namespace Gaffer;
+using namespace GafferScene;
+
+namespace
+{
+
+static InternedString g_lightsSetName( "__lights" );
+static InternedString g_defaultLightsSetName( "defaultLights" );
+static InternedString g_camerasSetName( "__cameras" );
+
+} // namespace
+
+GAFFER_NODE_DEFINE_TYPE( MotionPath );
+
+size_t MotionPath::g_firstPlugIndex = 0;
+
+MotionPath::MotionPath( const std::string &name )
+	:	FilteredSceneProcessor( name, PathMatcher::NoMatch )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	ValuePlugPtr startPlug = new ValuePlug( "start", Plug::In );
+	startPlug->addChild( new IntPlug( "mode", Plug::In, (int)FrameMode::Relative, /* min */ (int)FrameMode::Relative, /* max */ (int)FrameMode::Absolute ) );
+	startPlug->addChild( new FloatPlug( "frame", Plug::In, 0 ) );
+	addChild( startPlug );
+
+	ValuePlugPtr endPlug = new ValuePlug( "end", Plug::In );
+	endPlug->addChild( new IntPlug( "mode", Plug::In, (int)FrameMode::Relative, /* min */ (int)FrameMode::Relative, /* max */ (int)FrameMode::Absolute ) );
+	endPlug->addChild( new FloatPlug( "frame", Plug::In, 5 ) );
```

Comment:
I guess it depends on your intended use case, I don't think there is a great answer. My concern with a default trailing line is that normally scripts open on the first frame, so you would likely not see the curve until you scrub forward a bit, which could be confusing.

The 5 is totally random, just showed a reasonable line size in my example scenes, whereas 0,1 was often completely inside the geometry so again you couldn't see the curve at first.

Reply:
Agreed that there isn't a default that will work for everyone (or probably even ever), and that `+-5` magnitude is a is reasonable. I think I conceptualise this as "motion trails" though, I do definitely expect to see the trail behind the object by default.

Reply:
Fair enough, I don't really mind to be honest. I had been thinking of it like a trail myself until I decided the preview is more interesting (I can change the current xform interactively and see how it affects where I'm heading).

FWIW, someone just mentioned that since we default to centered motion blur, maybe a centered motion path would make sense (eg -2,2)?

Reply:
I don't see how center-on-frame blur relates to the style of the motion path you'd want at all. But I'll buy a centered path as a reasonable default on always-being-able-to-see-a-trail-of-some-sort grounds.

Reply:
Addressed in b1f88f60ce855b684faf2152319a08cfaf24f1c9 (Dan voted for centered default)